core/src/main/java/io/questdb/griffin/engine/functions/groupby/FirstDoubleGroupByFunction.java

```diff
@@ -74,18 +76,35 @@ public int getValueIndex() {
 
     @Override
     public boolean isParallelismSupported() {
-        return false;
+        return UnaryFunction.super.isParallelismSupported();
+    }
+
+    @Override
+    public boolean isReadThreadSafe() {
+        return UnaryFunction.super.isReadThreadSafe();
+    }
+
+    @Override
+    public void merge(MapValue destValue, MapValue srcValue) {
+        long srcRowId = srcValue.getLong(valueIndex);
+        long destRowId = destValue.getLong(valueIndex);
+        if (srcRowId != Numbers.LONG_NaN && (srcRowId < destRowId || destRowId == Numbers.LONG_NaN)) {
+            destValue.putLong(valueIndex, srcRowId);
+            destValue.putDouble(valueIndex + 1, srcValue.getDouble(valueIndex + 1));
+        }
     }
 
     @Override
     public void pushValueTypes(ArrayColumnTypes columnTypes) {
         this.valueIndex = columnTypes.getColumnCount();
-        columnTypes.add(ColumnType.DOUBLE);
+        columnTypes.add(ColumnType.LONG);   // row id
+        columnTypes.add(ColumnType.DOUBLE); // value
     }
 
     @Override
     public void setDouble(MapValue mapValue, double value) {
-        mapValue.putDouble(valueIndex, value);
+        mapValue.putLong(valueIndex, Numbers.LONG_NaN);
```

Comment:
it would be good to document (comments) these tricks with NaN

Reply:
Added comments in bdabd26b047ddb61a766eae93166770d53e564f8